CONTRIBUTING.md

```diff
@@ -207,6 +207,96 @@ When updating dependencies, please only update the particular dependencies you
+
+- **Debug**: Should primarily target consensus engine developers (i.e. core team
+  members and developers working on CometBFT forks).
+- **Info** and **Error**: Should primarily target operators and application
+  developers.
+
+### Log messages
+
+Log messages should always be tailored to the intended target audience. Unlike
+Go errors, **log messages must have the first letter of the message
+capitalized**. Only _errors_ in Go should start with a lowercase letter because
+they may end up being chained/embedded, but log messages are not chained in this
+same way.
+
+### Logging parameters
+
+In general, log messages should contain the **bare minimum** amount of
+information for those messages to be actionable by the target audience. So
+instead of dumping large quantities of raw data into the logs (e.g. RPC
+responses, transactions, block data), include lightweight references to the data
+that users can go look up via the RPC, CLI tools, etc. if they are genuinely
+interested in the details.
+
+When outputting variables, also keep in mind **concurrency concerns** of doing
+so. If outputting a pointer, understand that the value associated with that
+pointer at the time of calling the log function may differ to its value at the
+time it is finally serialized into the log message. Perhaps consider creating a
+temporary copy of the specific value you want to output and logging that
+temporary value. Also keep in mind potential data races when doing so.
+
+Finally, **use expensive operations like `fmt.Sprintf` sparingly**, as this can
+have a meaningful performance impact on a running production system. Consider
+and example where one may call `fmt.Sprintf` when logging something at **debug**
+level: even though an operator has configured their system to only log at
+**info** level and above, the expensive `fmt.Sprintf` calls will still take
+place, potentially slowing the system down. In such instances, consider printing
+values [lazily][log-lazy].
+
+### Examples of good log messages
+
+```golang
+// Operators generally wouldn't care whether an internal construct, like module
+// construction, has executed successfully.
+logger.Debug("Starting reactor", "module", "consensus")
+
+logger.Info("Committed block", "height", height, "appHash", appHash)
+
+// Include information about the error.
+logger.Error("Failed to execute important operation", "err", err)
+```
+
+### Examples of bad log messages
+
+```golang
+// Message starts with a lowercase letter, and will probably be called very
+// frequently, effectively spamming operators.
+logger.Debug("connected to peer", "peerID", peerID)
+
+// Potentially prints huge quantities of data (unnecessary) in a single message,
+// and at info level, spamming operators.
+logger.Info("Committed block", "block", fmt.Sprintf("%v", block))
+
+// Just as bad as the info-level message above because the (expensive)
+// fmt.Sprintf is always called, regardless of the operator's configured log
```

Comment:
Same comment as above `Debug` for `operators`

Reply:
This still applies. As per the comment, it still causes a performance hit _regardless_ of the operator's configured log level.